docs/public/operators-on-minikube.adoc

```diff
@@ -0,0 +1,204 @@
+
+[NOTE]
+====
+You can't always install an Operator in the `operators` namespace and expect it to be available across all namespaces. The Operator you're trying to installing needs to be designed to be available in this way as well. Certain Operators only support installation in a single namespace.
+
+Discussing this topic is out of scope of this guide so we have stated it as a note.
+====
+
+Enabling the `olm` addon will, among other things, create an `OperatorGroup` in the `operators` namepsace. Make sure that it's there:
+[source,sh]
+----
+$ kubectl get og -n operators
+NAME               AGE
+global-operators   3m37s
+----
+
+If you don't see one, create it using below command:
+[source,sh]
+----
+$ kubectl create -f - << EOF
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
+metadata:
+  name: global-operators 
+  namespace: operators 
+spec:
+  targetNamespaces:
+  - operators
+EOF
+----
+
+Now, install the etcd Operator using below command:
+[source,sh]
+----
+$ kubectl create -f - << EOF
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+metadata:
+  name: etcd
+  namespace: operators
+spec:
+  channel: clusterwide-alpha
+  name: etcd
+  source: operatorhubio-catalog
+  sourceNamespace: olm
+  startingCSV: etcdoperator.v0.9.4-clusterwide
+  installPlanApproval: Automatic
+EOF
+----
+
+Give it a few seconds before checking the availability of the etcd Operator:
+[source,sh]
+----
+$ odo catalog list services
+Operators available in the cluster
+NAME                                CRDs
+etcdoperator.v0.9.4-clusterwide     EtcdCluster, EtcdBackup, EtcdRestore
+----
+
+==== Troubleshooting
+
+If you don't see etcd Operator using above command or by doing `kubectl get csv -n operators`, make sure that pod belonging to the `CatalogSource` named `operatorhubio-catalog` is running:
+[source,sh]
+----
+$ kubectl get po -n olm | grep operatorhubio-catalog
+----
+
+If the state of this pod is `CrashLoopBackOff`, delete it so that Kubernetes will automatically spin up a new pod for the `CatalogSource`:
+
+[source,sh]
+----
+$ kubectl delete po -n olm <name-of-operatorhubio-catalog-pod>
+----
+
+Once the pod for this `CatalogSource` is up, wait a few seconds before trying to find the etcd Operator when you do `odo catalog list services`.
+
+=== Installing the Service Binding Operator
+
+We use the link:https://github.com/redhat-developer/service-binding-operator/[Service Binding Operator] to provide `odo link` feature which links an odo component with an Operator backed service. Thus, to be able to use this feature, it's essential that we install the Operator first.
+
+Service Binding Operator is not yet available via the OLM. The team is link:https://github.com/redhat-developer/service-binding-operator/issues/727[working on making it available] through OLM.
```

Comment:
Actually it is available, but not in upstream index, instead in `registry.redhat.io/redhat/community-operator-index:latest`

Hence, we could define the following catalogsource:

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: community-operators
  namespace: operators
spec:
  displayName: Community Operators
  image: registry.redhat.io/redhat/community-operator-index:latest
  publisher: Red Hat
  sourceType: grpc
  priority: -400
  publisher: Red Hat
  updateStrategy:
    registryPoll:
      interval: 10m0s
```

And then add subscription:

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: service-binding-operator
  namespace: operators
spec:
  channel: 'beta'
  installPlanApproval: Automatic
  name: service-binding-operator
  source: community-operators
  sourceNamespace: operators
```

That should install the operator.

Reply:
I tried this on a fresh minikube cluster after enabling olm - `minikube addons enable olm`. But the Opertor doesn't show up. Here's `kubectl describe` for the catsrc:
```
Name:         community-operators
Namespace:    operators
Labels:       <none>
Annotations:  API Version:  operators.coreos.com/v1alpha1
Kind:         CatalogSource
Metadata:
  Creation Timestamp:  2020-10-28T06:00:19Z
  Generation:          1
  Managed Fields:
    API Version:  operators.coreos.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:displayName:
        f:image:
        f:publisher:
        f:sourceType:
        f:updateStrategy:
          .:
          f:registryPoll:
            .:
            f:interval:
    Manager:      kubectl
    Operation:    Update
    Time:         2020-10-28T06:00:19Z
    API Version:  operators.coreos.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        f:icon:
          .:
          f:base64data:
          f:mediatype:
      f:status:
        .:
        f:connectionState:
          .:
          f:address:
          f:lastConnect:
          f:lastObservedState:
        f:registryService:
          .:
          f:createdAt:
          f:port:
          f:protocol:
          f:serviceName:
          f:serviceNamespace:
    Manager:         catalog
    Operation:       Update
    Time:            2020-10-28T06:06:48Z
  Resource Version:  1614
  Self Link:         /apis/operators.coreos.com/v1alpha1/namespaces/operators/catalogsources/community-operators
  UID:               5b722962-5886-4295-a165-59d7b79d2d8d
Spec:
  Display Name:  Community Operators
  Image:         registry.redhat.io/redhat/community-operator-index:latest
  Publisher:     Red Hat
  Source Type:   grpc
  Update Strategy:
    Registry Poll:
      Interval:  10m0s
Status:
  Connection State:
    Address:              community-operators.operators.svc:50051
    Last Connect:         2020-10-28T06:06:48Z
    Last Observed State:  TRANSIENT_FAILURE
  Registry Service:
    Created At:         2020-10-28T06:00:19Z
    Port:               50051
    Protocol:           grpc
    Service Name:       community-operators
    Service Namespace:  operators
Events:                 <none>
```
I think the issue with pulling the image:
```
$ docker pull registry.redhat.io/redhat/community-operator-index:latest
Error response from daemon: Get https://registry.redhat.io/v2/redhat/community-operator-index/manifests/latest: unauthorized: Please login to the Red Hat Registry using your Customer Portal credentials. Further instructions can be found here: https://access.redhat.com/RegistryAuthentication
```
Is it possible to host the image on quay and provide that instead?

Reply:
Indeed @dharmit  @pedjak we get an ImagePullBackOff err
`operators     community-operators-6r6q4                   0/1     ImagePullBackOff   0          110s
`

Reply:
Yeah, we need to add credentials for that repo into the cluster.

Here is what I did:

* Follow https://access.redhat.com/RegistryAuthentication and create account if you do not have any
* Verify that your credentials are correct using `docker login -u <your_username> -p <you_passwd> registry.redhat.io`
* start minikube cluster
* enable OLM with `minikube addons enable olm` or install it with `curl -sL https://github.com/operator-framework/operator-lifecycle-manager/releases/download/0.16.1/install.sh | bash -s 0.16.1`
* create a new image pull secret out of your local `.docker/config.json` file:

```
kubectl create secret generic community-operators-secrets -n olm --from-file=.dockerconfigjson=~/.docker/config.json --type=kubernetes.io/dockerconfigjson
```
* add that pull secret to the default account in `olm` namespace

```
kubectl patch serviceaccount default -p '{"imagePullSecrets": [{"name": "community-operators-secrets"}]}'
```

* add catalog source to `olm` namespace

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: community-operators
  namespace: olm
spec:
  displayName: Community Operators
  image: registry.redhat.io/redhat/community-operator-index:latest
  publisher: Red Hat
  sourceType: grpc
  publisher: Red Hat
  updateStrategy:
    registryPoll:
      interval: 10m0s
```
* eventually `kubectl get packagemanifest -n olm` will list service-binding-operator as well
* add subscription to `operators` namespace:

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: service-binding-operator
  namespace: operators
spec:
  channel: 'beta'
  installPlanApproval: Automatic
  name: service-binding-operator
  source: community-operators
  sourceNamespace: olm
```

The operator will eventually be up and running

Reply:
@pedjak I'm making this comment without validating above steps because I'm quite sure that it will work. However, the point is to be able to bring down the barrier to use SBO on minikube/k8s environments.

I'm not sure if users would appreciate the idea of having to create an account on RH Customer Portal to just use an Operator. If we could manage to address https://github.com/redhat-developer/service-binding-operator/issues/727 and make SBO easily available, that would really help.

If I see this as an end user (an enterprise developer who only cares about writing java/nodejs apps that connect to an external service), there's a tonne of stuff I need to do to run Operators and link them with an odo component (`Deployment` in k8s/OCP lingo).

@kadel @sbose78 WDYT?

Reply:
@dharmit we definitelly have a plan to publish the operator on upstream operatorhub - creating account is just a temp thing until this does not happen.

Reply:
Most of the upstream Kubernetes won't have RH account. We want to be upstream friendly and RH account requirement is blocker for upstream docs.

Reply:
> Most of the upstream Kubernetes won't have RH account. We want to be upstream friendly and RH account requirement is blocker for upstream docs.

Sure @kadel - I cannot agree more. I have demonstrated above only what is currently available, given that we still do not publish to upstream operatorhub. However, that's our near term plan, but in order that this can happen, we need to first verify if our acceptance tests are green on plain k8s clusters. cc @pmacik